java/clients/src/main/java/sleeper/clients/teardown/TerminateEMRClusters.java

```diff
@@ -0,0 +1,115 @@
+
+import com.amazonaws.services.elasticmapreduce.AmazonElasticMapReduce;
+import com.amazonaws.services.elasticmapreduce.model.ClusterState;
+import com.amazonaws.services.elasticmapreduce.model.ClusterSummary;
+import com.amazonaws.services.elasticmapreduce.model.ListClustersRequest;
+import com.amazonaws.services.elasticmapreduce.model.TerminateJobFlowsRequest;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import sleeper.configuration.properties.InstanceProperties;
+import sleeper.util.PollWithRetries;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+import static java.lang.Math.min;
+import static sleeper.configuration.properties.UserDefinedInstanceProperty.ID;
+
+public class TerminateEMRClusters {
+    private static final Logger LOGGER = LoggerFactory.getLogger(TerminateEMRClusters.class);
+    private static final long POLL_INTERVAL_MILLIS = 30000;
+    private static final int MAX_POLLS = 30;
+
+    private final PollWithRetries poll = PollWithRetries.intervalAndMaxPolls(POLL_INTERVAL_MILLIS, MAX_POLLS);
+
+    private final AmazonElasticMapReduce emrClient;
+    private final String clusterPrefix;
+    private static final List<String> ACTIVE_STATES = List.of(ClusterState.STARTING.name(), ClusterState.BOOTSTRAPPING.name(),
+            ClusterState.RUNNING.name(), ClusterState.WAITING.name(), ClusterState.TERMINATING.name());
+
+    public TerminateEMRClusters(AmazonElasticMapReduce emrClient, InstanceProperties properties) {
+        this.emrClient = emrClient;
+        this.clusterPrefix = "sleeper-" + properties.get(ID) + "-";
+    }
+
+    public void run() throws InterruptedException {
+        List<ClusterSummary> clusters = emrClient.listClusters(new ListClustersRequest()
+                .withClusterStates(ACTIVE_STATES)).getClusters();
+        List<String> clusterIds = clusters.stream()
+                .filter(cluster -> cluster.getName().startsWith(clusterPrefix))
+                .map(ClusterSummary::getId)
+                .collect(Collectors.toList());
+        if (clusterIds.size() == 0) {
+            LOGGER.info("No running clusters to terminate");
+        } else {
+            LOGGER.info("Terminating {} running clusters", clusterIds.size());
+            terminateClusters(clusterIds);
+            LOGGER.info("Waiting for clusters to terminate");
+            pollUntilTerminated();
+        }
+    }
+
+    private void terminateClusters(List<String> clusters) {
+        // Can only terminate 10 clusters at a time
+        // See https://docs.aws.amazon.com/emr/latest/APIReference/API_TerminateJobFlows.html
+        for (int i = 0; i < clusters.size(); i += 10) {
+            List<String> clusterBatch = clusters.subList(i, min(i + 10, clusters.size()));
+            emrClient.terminateJobFlows(new TerminateJobFlowsRequest().withJobFlowIds(clusterBatch));
+            LOGGER.info("Terminated {} clusters out of {}", min(i + 10, clusters.size()), clusters.size());
```

Comment:
It might be worth setting a local variable for the end index rather than calculating it twice.